src/platforms/android/androidiaphandler.cpp

```diff
@@ -0,0 +1,308 @@
+void AndroidIAPHandler::onSkuDetailsReceived(JNIEnv* env, jobject thiz,
+                                             jstring data) {
+  Q_UNUSED(thiz);
+
+  const char* buffer = env->GetStringUTFChars(data, nullptr);
+  if (!buffer) {
+    logger.error() << "onSkuDetailsReceived - failed to parse data.";
+    return;
+  }
+  QByteArray raw = QByteArray(buffer);
+  env->ReleaseStringUTFChars(data, buffer);
+
+  logger.debug() << "onSkuDetailsReceived - parsing raw data: " << raw;
+
+  QJsonParseError jsonError;
+  QJsonDocument json = QJsonDocument::fromJson(raw, &jsonError);
+
+  if (QJsonParseError::NoError != jsonError.error) {
+    logger.error() << "onSkuDetailsRecieved, Error parsing json. Code: "
+                   << jsonError.error << "Offset: " << jsonError.offset
+                   << "Message: " << jsonError.errorString() << "Data: " << raw;
+    return;
+  }
+
+  if (!json.isObject()) {
+    logger.error() << "onSkuDetailsReceived - object expected.";
+    return;
+  }
+
+  QJsonObject obj = json.object();
+  if (!obj.contains("products")) {
+    logger.error() << "onSkuDetailsReceived - products entry expected.";
+    return;
+  }
+
+  QJsonArray products = obj["products"].toArray();
+  if (products.isEmpty()) {
+    logger.error() << "onSkuDetailsRecieved - no products found.";
+    return;
+  }
+  IAPHandler* iap = IAPHandler::instance();
+  static_cast<AndroidIAPHandler*>(iap)->updateProductsInfo(products);
+  iap->productsRegistrationCompleted();
+}
+
+void AndroidIAPHandler::updateProductsInfo(const QJsonArray& returnedProducts) {
+  Q_ASSERT(m_productsRegistrationState == eRegistering);
+
+  QList<QString> productsUpdated;
+  for (auto product : returnedProducts) {
+    QString productIdentifier = product["sku"].toString();
+    Product* productData = findProduct(productIdentifier);
+    Q_ASSERT(productData);
+
+    productData->m_price = product["totalPriceString"].toString();
+    productData->m_monthlyPrice = product["monthlyPriceString"].toString();
+    productData->m_nonLocalizedMonthlyPrice =
+        product["monthlyPrice"].toDouble();
+
+    productsUpdated.append(productIdentifier);
+  }
+  // Remove products from m_products if we didn't get info back from google
+  // about them.
+  for (auto product : m_products) {
+    if (!productsUpdated.contains(product.m_name)) {
+      unknownProductRegistered(product.m_name);
+    }
+  }
+}
+
+void AndroidIAPHandler::nativeStartSubscription(Product* product) {
+  auto jniString = QAndroidJniObject::fromString(product->m_name);
+  auto appActivity = QtAndroid::androidActivity();
+
+  QAndroidJniObject::callStaticMethod<void>(
+      "org/mozilla/firefox/vpn/InAppPurchase", "purchaseProduct",
+      "(Ljava/lang/String;Landroid/app/Activity;)V", jniString.object(),
+      appActivity.object());
+}
+
+// static
+void AndroidIAPHandler::onNoPurchases(JNIEnv* env, jobject thiz) {
+  Q_UNUSED(env)
+  Q_UNUSED(thiz);
+  // ToDo - I'm not sure when this scenario would occur
```

Comment:
Ok... this is a silly comment, but, if we use "TODO", my vim shows them highlighted ... :)

Reply:
I'll try and remember :D